File: 1/2020.10.12/64.Minimum Path Sum/python/solution.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        # dp, use grid itself
        if not grid or not grid[0]:
            return 0
        m, n = len(grid), len(grid[0])
        for i in range(1, n):
            grid[0][i] += grid[0][i-1]
        for i in range(1, m):
            grid[i][0] += grid[i-1][0]
        for i in range(1, m):
            for j in range(1, n):
                grid[i][j] += min(grid[i-1][j], grid[i][j-1])
        return grid[-1][-1]

    def minPathSum(self, grid: List[List[int]]) -> int:
        # dp, use grid itself
        if not grid or not grid[0]:
            return 0
        m, n = len(grid), len(grid[0])
        dp = [None for _ in range(n)]
        for i in range(n):
            if i == 0:
                dp[i] = grid[0][i]
            else:
                dp[i] = grid[0][i] + dp[i-1]
        for i in range(1, m):
            for j in range(n):
                if j == 0:
                    dp[j] += grid[i][j]
                else:
                    dp[j] = grid[i][j] + min(dp[j], dp[j-1])
        return dp[-1]

File: 1/2020.10.12/64.Minimum Path Sum/java/solution.java
class Solution {
    public int minPathSum(int[][] grid) {
        // dp, use grid itself
        if (grid.length == 0 || grid[0].length == 0) {
            return 0;
        }
        int m = grid.length, n = grid[0].length;
        // first row
        for (int i = 1; i < n; i++) {
            grid[0][i] += grid[0][i-1];
        }
        // first column
        for (int i = 1; i < m; i++) {
            grid[i][0] += grid[i-1][0];
        }
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                grid[i][j] += Math.min(grid[i-1][j], grid[i][j-1]);
            }
        }
        return grid[m-1][n-1];
    }

    public int minPathSum(int[][] grid) {
        // dp, O(n) space
        if (grid.length == 0 || grid[0].length == 0) {
            return 0;
        }
        int m = grid.length, n = grid[0].length;
        int[] dp = new int[n];
        dp[0] = grid[0][0];
        for (int i = 1; i < n; i++) {
            dp[i] = grid[0][i] + dp[i-1];
        }
        for (int i = 1; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (j == 0) {
                    dp[j] += grid[i][j];
                } else {
                    dp[j] = grid[i][j] + Math.min(dp[j], dp[j-1]);
                }
            }
        }
        return dp[n-1];
    }
}
